src/common/offline_transformations/src/pruning/propagate_masks.cpp

```diff
@@ -1420,6 +1422,196 @@ class ngraph::pass::mask_propagation::Transpose : public MatcherPass {
+
+        ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher& m) {
+            const auto& pattern_map = m.get_pattern_value_map();
+            auto axis_node = ngraph::as_type<opset6::Constant>(pattern_map.at(axis_pattern).get_node());
+            if (!axis_node)
+                return false;
+
+            const auto& input = pattern_map.at(input_pattern);
+            const auto input_mask = getMask(input);
+            if (!input_mask)
+                return false;
+
+            auto axis = axis_node->cast_vector<int64_t>()[0];
+            if (axis < 0)
+                axis += input_mask->size();
+
+            auto split = pattern_map.at(split_pattern).get_node();
+
+            auto split_lengths_const =
+                ngraph::as_type<opset6::Constant>(pattern_map.at(split_lengths_pattern).get_node());
+            if (!split_lengths_const)
+                return false;
+            auto split_lengths = split_lengths_const->cast_vector<int64_t>();
+
+            // adjust split_lengths if needed
+            // split_lengths can contain -1 value
+            int minus_one_length_idx = -1;
+            int64_t total_lengths = 0;
+            for (int i = 0; i < split_lengths.size(); i++) {
+                if (split_lengths[i] == -1) {
+                    minus_one_length_idx = i;
+                    continue;
+                }
+                total_lengths += split_lengths[i];
+            }
+            if (minus_one_length_idx >= 0 && (*input_mask)[axis].size() > 0) {
+                const auto& input_shape = input.get_partial_shape();
+                if (input_shape[axis].is_dynamic())
+                    return false;
+                auto split_dim = static_cast<uint64_t>(input_shape[axis].get_length());
+                split_lengths[minus_one_length_idx] = split_dim - total_lengths;
+            }
+
+            uint64_t split_start = 0;
+            uint64_t split_end = split_lengths[0];
+            std::vector<Mask::Ptr> output_masks;
+            std::vector<Mask*> output_masks_raw;
+            Mask* input_mask_raw = input_mask.get();
+            for (size_t i = 0; i < split->get_output_size(); i++) {
+                auto output_mask = std::make_shared<Mask>();
+                output_mask->copy_and_slice_mask_from(input_mask_raw, axis, split_start, split_end);
+                output_masks.push_back(output_mask);
+                output_masks_raw.push_back(output_mask.get());
+                output_mask->add_callback(
+                    [input_mask_raw, axis, split_start, split_end](Mask::Ptr cur_mask) -> bool {
+                        cur_mask->copy_and_slice_mask_from(input_mask_raw, axis, split_start, split_end);
+                        return true;
+                    },
+                    input_mask);
+                split_start = split_end;
+                if (i + 1 < split_lengths.size())
+                    split_end += split_lengths[i + 1];
```

Comment:
Please use one offset variable like `split_offset`, I think it would be more clear